File: lib/plot_cnv.R

# get passed arguments
plotFile     <- Sys.getenv("plotFile")
spanFile     <- Sys.getenv("spanFile")
snpFile      <- Sys.getenv("snpFile")
gapDupFile   <- Sys.getenv("gapDupFile")
plotDir      <- Sys.getenv("plotDir")
centerWidth  <- as.numeric(Sys.getenv("centerWidth"))
maxSnpCount  <- as.numeric(Sys.getenv("maxSnpCount"))
refName      <- Sys.getenv("refName")
altName      <- Sys.getenv("altName")

# set variables
widthPadding <- (centerWidth - 1) / 2;
idx_col <- 'darkred'
oth_col <- 'blue'
non_col <- 'grey'
cnlim   <- c(-2,4)
snplim  <- c(-maxSnpCount,maxSnpCount)

# get the data
write("  loading data", file=stderr())
d <- read.table(plotFile, header=TRUE, sep="\t", stringsAsFactors=FALSE)
colnames(d) <- c('chrom', 'region', 'group', 'grpId', 'cnvType', 'cnvSize', 'cnvFrac',
                 'idx_sample', 'oth_sample',
                 'start', 'end', 'left', 'right',
                 'sample', 'center', 'modCN', 'dCN');
s <- read.table(spanFile, header=TRUE, sep="\t", stringsAsFactors=FALSE)
colnames(s) <- c('grpId', 'idx_sample', 'sample', 'cnvType', 'start', 'end', 'row');
S <- read.table(snpFile, header=TRUE, sep="\t", stringsAsFactors=FALSE)
colnames(S) <- c('grpId', 'idx_sample', 'sample', 'pos', 'nRef', 'nAlt', 'refBase', 'altBase');
g <- read.table(gapDupFile, header=TRUE, sep="\t", stringsAsFactors=FALSE)
colnames(g) <- c('grpId', 'start', 'end', 'col');

# plot functions
addVLines <- function(){
    abline(v=c(left, right)/1e6, col="grey", lty=2)
}
addGaps <- function(lim){
    if(nrow(gg) > 0){
        for(i in 1:nrow(gg)){
            rect(gg[i,'start']/1e6, lim[1], gg[i,'end']/1e6, lim[2], col=gg[i,'col'], border=NA)
        }
    }
}
plotCopyNumber <- function(){
    legend("topleft", c('Modal Copy Number', idx_sample), col=c('black', idx_col), lwd=1, bty="n", cex=1)
    addGaps(cnlim)
    abline(h=seq(-2,4,1), col="grey", lty=2)    
    abline(h=0, col="black", lty=1)
    addVLines()
    isFirst <- TRUE
    for (sample in unique(ddd[,'sample'])){    
        dddd <- ddd[ddd$sample==sample,]
        if(nrow(dddd)>0){
            if(isFirst){
                dddd[,'modCN'] <- ifelse(dddd$modCN < cnlim[1], cnlim[1], dddd$modCN)
                dddd[,'modCN'] <- ifelse(dddd$modCN > cnlim[2], cnlim[2], dddd$modCN) 
                lines (dddd$center/1e6, dddd$modCN, col="black")
                isFirst <- FALSE
            }                 
            dddd[,'dCN'] <- ifelse(dddd$dCN < cnlim[1], cnlim[1], dddd$dCN)        
            dddd[,'dCN'] <- ifelse(dddd$dCN > cnlim[2], cnlim[2], dddd$dCN)
            if(sample==idx_sample) {   
                lines (dddd$center/1e6, dddd$dCN, col=idx_col)
                points(dddd$center/1e6, dddd$dCN, col=idx_col, pch=20)
            } else if(sample==oth_sample) {
                lines (dddd$center/1e6, dddd$dCN, col=oth_col)   
            } else {
                lines (dddd$center/1e6, dddd$dCN, col=non_col)
            }
        }
    }  
}
getBias <- function(snps){
    nRef <- sum(snps$nRef)
    nAlt <- sum(snps$nAlt)
    nReads <- nRef + nAlt
    ifelse(nReads>0, (nRef - nAlt) / nReads * snplim[2], 0)
}
plotSnpCounts <- function(){
    addGaps(snplim)
    abline(h=((2 - 1) / 3 * snplim[2]), col="grey", lty=1)
    abline(h=((1 - 2) / 3 * snplim[2]), col="grey", lty=1)
    abline(h=((1 - 0) / 1 * snplim[2]), col="grey", lty=1)
    abline(h=((0 - 1) / 1 * snplim[2]), col="grey", lty=1)
    abline(h=0, col="black", lty=1)
    addVLines()
    mtext(refName, side=2, line=2, at=snplim[2]*0.9, las=1, cex=0.8)
    mtext(altName, side=2, line=2, at=snplim[1]*0.9, las=1, cex=0.8)
    SSS <- SS[SS$sample==idx_sample,]    
    if(nrow(SSS)>0){
        nRef <- ifelse(SSS$nRef > snplim[2],  snplim[2],  SSS$nRef)                 
        nAlt <- ifelse(SSS$nAlt > snplim[2], -snplim[2], -SSS$nAlt)        
        points(SSS$pos/1e6, nRef, col="black", type="h")
        points(SSS$pos/1e6, nAlt, col="black", type="h")        
    }
    for(sample in c(oth_sample, idx_sample)){
        SSS <- SS[SS$sample==sample,]
        if(nrow(SSS)>0){
            cnvBias <- getBias(SSS[SSS$pos>=left&SSS$pos<=right,])
            lftBias <- getBias(SSS[SSS$pos>=start&SSS$pos<left,])
            rgtBias <- getBias(SSS[SSS$pos>right&SSS$pos<=end,])
            col <- if(sample==oth_sample) {oth_col} else {idx_col}
            lines(c(left,  right) / 1e6, rep(cnvBias,2), col=col, lwd=2)
            lines(c(start, left)  / 1e6, rep(lftBias,2), col=col, lwd=2)
            lines(c(right, end)   / 1e6, rep(rgtBias,2), col=col, lwd=2)                 
        }
    }   
}
plotSpans <- function(sample, col){
    legend("topleft", sample, col=col, lwd=1, bty="n", cex=1)
    addGaps(ylim)
    addVLines()    
    sss <- ss[ss$sample==sample,]
    if(nrow(sss)>0){
        segments(sss$start/1e6, sss$row, sss$end/1e6, sss$row, col=col, lwd=1, lend=2)
        abline(h=0.5, col="grey", lty=2)
        for(cnvType in c('del', 'dup', 'invF', 'invR')){
            ssss <- sss[sss$cnvType==cnvType,]
            if(nrow(ssss)>0){
                min <- min(ssss[,'row'])
                max <- max(ssss[,'row'])
                abline(h=max+0.5, col="grey", lty=2)
                mtext(cnvType, side=2, line=1, at=min+(max-min)/2, las=1, cex=0.8)                
            }    
        }   
    }
}

# make the plots
for (grpId in unique(c(d$grpId, s$grpId))){
    
    # get this index event
    write(paste("  ", grpId, sep=""), file=stderr())    
    dd      <- d[d$grpId==grpId,]
    ss      <- s[s$grpId==grpId,]
    SS      <- S[S$grpId==grpId,]
    gg      <- g[g$grpId==grpId,]
    group   <- dd[1,'group']
    
    # set the coordinates
    chrom   <- dd[1,'chrom']
    start   <- dd[1,'start']
    end     <- dd[1,'end']
    left    <- dd[1,'left']
    right   <- dd[1,'right']
    clim    <- c(start, end) /  1e6

    # determine the index and other samples
    idx_samples <- unique(c(dd$idx_sample, ss$idx_sample)) # can be more than one if region has multiple candidate samples
    idx_sample  <- idx_samples[1]
    ddd <- dd[dd$idx_sample==idx_sample,] # only need to plot the trace from one call
    oth <- ddd[ddd$sample!=idx_sample&ddd$center>=left&ddd$center<=right,]    
    dlt <- abs(oth[,'dCN'])    
    oth_sample <- if(!is.na(ddd[1,'oth_sample'])){
        ddd[1,'oth_sample']
    } else if(length(dlt)>0){
        oth[which(dlt==max(dlt))[1],'sample']
    } else {
        'xxxx'
    }

    # open the jpg file
    jpgFile <- paste(ddd[1,'region'], group, grpId, "jpg", sep=".")
    jpgFile <- paste(plotDir, jpgFile, sep="/")        
    jpeg(jpgFile,
         width=5, height=6, units="in", pointsize=9,
         res=400, quality=80)    
    
    # set the composite layout
    layout(matrix(c(1,2,3,4), 4, 1, byrow=TRUE), heights=c(0.28,0.22,0.22,0.28))

    # plot the copy number traces
    par(mar=c(0.1, 4.1, 4.1, 0.5))
    plot(0, 0, type="n", xlim=clim, ylim=cnlim,
         xaxt="n",
         ylab="Copy Number (Change)",
         main=paste(paste(group, " #", grpId, ", ",
                          ddd[1,'cnvType'], " ", as.numeric(ddd[1,'cnvSize']), " bp", sep=""))
    )
    plotCopyNumber()

    # plot the SNP counts
    par(mar=c(0.1, 4.1, 0.1, 0.5))    
    plot(0, 0, type="n", xlim=clim, ylim=snplim,
         xaxt="n",
         ylab="SNP Count"
    )
    plotSnpCounts()

    # plot the index sample spans
    par(mar=c(0.1, 4.1, 0.1, 0.5))
    rows <- ss[ss$sample==idx_sample,'row']
    ylim <- c(0, 1 + if(length(rows)>0) {max(rows)} else {0})
    plot(0, 0, type="n", xlim=clim, ylim=ylim,
         xaxt="n",
         yaxt="n", ylab=""
    )
    plotSpans(idx_sample, idx_col)
    
    # plot the other/control sample spans
    par(mar=c(4.1, 4.1, 0.1, 0.5))
    rows <- ss[ss$sample==oth_sample,'row']
    ylim <- c(0, 1 + if(length(rows)>0) {max(rows)} else {0})
    plot(0, 0, type="n", xlim=clim, ylim=ylim,
         xlab=paste(chrom, "coordinate (Mb)", sep=" "),
         yaxt="n", ylab=""
    )
    plotSpans(oth_sample, oth_col)

    # finish the jpg
    graphics.off()
}

File: lib/plot_cnv.pl
use strict;
use warnings;

# define variables
use vars qw($version $utility $error $libDir
            $TRUE $FALSE
            $tmpH 
            $inputDir $tmpDir $outputDir $plotDir
            %pPrpCols
            %binFileCols $nbinFileLeadCol %binColOffsets %gntCols);
my %c = %pPrpCols;
my $bsCol = $binFileCols{start};
my $beCol = $binFileCols{end};
my $cnCol = $binFileCols{cnRaw};
my ($group, $callBinsFile, $idxAnomFile, $genSnpsDir,
    $centerWidth,
    $gapFile, $segDupFile, $maxSnpCount,
    $refName, $altName,
    $widthPadding, $bamGlobXX, $minQualXX);
my ($callGroup, $callSmpFile, %samples, $nSamples, @samples, %allSamples);
my ($posH, $datH, $spnH, $snpH, $prpH, $lstH, $bamH, $gapH);

# manage options
sub setOptions_plot_cnv {
    setOptionValue(\$callBinsFile,  'call-bins-file');
    setOptionValue(\$idxAnomFile,   'idx-anom-file');
    setOptionValue(\$genSnpsDir,    'gen-snps-dir');
    setOptionValue(\$group,         'group');
    setOptionValue(\$plotDir,       'plot-dir');
    setOptionValue(\$tmpDir,        'tmp-dir',      '/tmp');
    setOptionValue(\$centerWidth,   'center-width', 5);
    setOptionValue(\$gapFile,       'gap-file');
    setOptionValue(\$segDupFile,    'seg-dup-file');
    setOptionValue(\$maxSnpCount,   'max-snp-count', 5);
    setOptionValue(\$refName,       'ref-name', "Ref");
    setOptionValue(\$altName,       'alt-name', "Alt");
    #------------------------------------------------------
    -e $callBinsFile or die "$error: $callBinsFile not found\n";
    $callBinsFile =~ m/svtools\.call\.bins\.(.+)\.bgz$/ or die "not a call bins file: $callBinsFile\n";
    $callGroup = $1;
    $callSmpFile = $callBinsFile;
    $callSmpFile =~ s/bins\.$callGroup\.bgz/samples\.$callGroup\.txt/;
    -e $idxAnomFile or die "$error: $idxAnomFile not found\n";
    $idxAnomFile =~ m/svtools\.index\.anomalies\.(.+)\.bgz$/ or die "not an index anomalies file: $idxAnomFile\n";
    -d $plotDir or die "$error: $plotDir does not exist or is not a directory\n";
    -d $tmpDir  or die "$error: $tmpDir does not exist or is not a directory\n";
    -d $genSnpsDir or die "$error: $genSnpsDir does not exist or is not a directory\n";
    $centerWidth =~ m|\D| and die "$error: center-width must be an integer number\n";
    $gapFile and (-e $gapFile or die "$error: $gapFile not found\n");
    $segDupFile and (-e $segDupFile or die "$error: $segDupFile not found\n");
    $outputDir = $plotDir;
    $widthPadding = ($centerWidth - 1) / 2;
}

# main execution block
sub svtools_plot_cnv {

    # initialize
    setOptions_plot_cnv();    
    print STDERR "$utility plot_cnv: " . getTime(), "\n";
    openInputStream($callSmpFile, \$tmpH);
    my $samples = <$tmpH>;
    closeHandles($tmpH);
    chomp $samples;
    @samples = split("\t", $samples);
    %samples = map { $samples[$_] => $_ } 0..$#samples;
    $nSamples = @samples;

    # collect data for cnv span on the input stream
    print STDERR "collecting data\n";
    my $plotFile = getTmpFile("plot_cnv", 'data', "gz");
    openOutputStream($plotFile, \$datH, $TRUE);
    my $spanFile = getTmpFile("plot_cnv", 'spans', "gz");
    openOutputStream($spanFile, \$spnH, $TRUE);
    my $snpFile = getTmpFile("plot_cnv", 'snps', "gz");
    openOutputStream($snpFile, \$snpH, $TRUE);
    my $gapDupFile  = getTmpFile("plot_cnv", 'gaps_dups', "gz");
    openOutputStream($gapDupFile, \$gapH, $TRUE);
    my $lstFile = getOutFile("list", $group, "txt");
    openOutputStream($lstFile, \$lstH, $FALSE);
    openInputStream(undef, \$prpH);    
    my $header = <$prpH>;
    print $lstH $header;
    my $nCnvs = 0;
    while (my $prp = <$prpH>) {
        $nCnvs++;
        print $lstH $prp;
        chomp $prp;
        my @prp = split("\t", $prp);
        my $width = $prp[$c{end}] - $prp[$c{start}] - 1;
        my $start = $prp[$c{start}] - $width * $widthPadding;
        my $end = $prp[$c{end}] + $width * $widthPadding;
        my $plotRegion = "$prp[$c{chrom}]:$start-$end";   
        $prp[$c{region}] =~ s/:/-/;
        print STDERR "  $prp[$c{grpId}]  = $plotRegion\n";
        printDepthBins(\@prp, $plotRegion, $start, $end, $prp[$c{start}], $prp[$c{end}]);
        printAnomSpans(\@prp, $plotRegion, $start, $end);
        printSNPs(\@prp, $plotRegion);
        printGaps(\@prp, $plotRegion);
    }
    closeHandles($datH, $spnH, $snpH, $prpH, $lstH, $gapH);

    # make the plots
    if ($nCnvs) {
        print STDERR "making plots\n";
        $ENV{plotFile}     = $plotFile;
        $ENV{spanFile}     = $spanFile;
        $ENV{snpFile}      = $snpFile;
        $ENV{gapDupFile}   = $gapDupFile;
        $ENV{plotDir}      = $plotDir;
        $ENV{centerWidth}  = $centerWidth;
        $ENV{maxSnpCount}  = $maxSnpCount;
        $ENV{refName}      = $refName;
        $ENV{altName}      = $altName;
        system("Rscript $libDir/plot_cnv.R");
    } else {
        print STDERR "nothing to plot\n";  
    }

    # finish up 
    unlink $plotFile;
    unlink $spanFile;
    unlink $snpFile;
    unlink $gapDupFile;
}

# get the data for the coverage depth traces
sub printDepthBins {
    my ($prp, $plotRegion, $start, $end, $left, $right) = @_;
    open my $tbxH, "-|", "tabix $callBinsFile $plotRegion" or
        die "could not open tabix stream: tabix $callBinsFile $plotRegion\n";
    while (my $bin = <$tbxH>) {
        chomp $bin;
        my @bin = split("\t", $bin);
        my $center = int($bin[$bsCol] + ($bin[$beCol] - $bin[$bsCol]) / 2 + 0.5);
        my $modCN  = $bin[$cnCol];
        foreach my $sampleI(0..$#samples){
            my $dCN = $bin[$nbinFileLeadCol + $sampleI + $nSamples * $binColOffsets{smpDCNRaw}];
            print $datH join("\t",
                (map {$$prp[$c{$_}]} qw(chrom region group grpId cnvType cnvSize cnvFrac sample othSample)),
                $start, $end, $left, $right,
                $samples[$sampleI], $center, $modCN, $dCN), "\n";            
        }        
    }
    close $tbxH;
}

# get the data for the anomalous fragment tracks
sub printAnomSpans {
    my ($prp, $plotRegion, $start, $end) = @_;
    
    # split by sample and pair type      
    open my $tbxH, "-|", "tabix $idxAnomFile $plotRegion" or
        die "could not open tabix stream: tabix $idxAnomFile $plotRegion\n";
    my %pairs;
    my @pairTypes = qw(del dup invF invR);    
    my %pairTypes = map { $_ => 1 } @pairTypes;
    %allSamples = map { $_ => 1 } @samples;
    while (my $bed = <$tbxH>) {
        chomp $bed;
        my @bed = split("\t", $bed);
        my $okSize = (($bed[2] - $bed[1]) < 2.5e6);
        my $inWindow = (($bed[1] >= $start and $bed[1] <= $end) and 
                        ($bed[2] >= $start and $bed[2] <= $end));
        my $oneEndIn = (($bed[1] >= $start and $bed[1] <= $end) or 
                        ($bed[2] >= $start and $bed[2] <= $end));
        ($inWindow or ($oneEndIn and $okSize)) or next;     
        $bed[3] =~ m/(.+)\..+,(\w+)/; # sample, cnvType        
        $pairTypes{$2} or next;
        push @{$pairs{$1}{$2}{$bed[1]}{$bed[2]}}, \@bed;
        $allSamples{$1}++;
    }    
    close $tbxH;
    
    # plot a track for each pair type
    foreach my $sample(keys %allSamples){
        my $maxLineAll = 1;
        foreach my $pairType(@pairTypes){
            my $bed = $pairs{$sample}{$pairType} or next;
            my ($pckBed, $maxLine) = packSpans($bed);
            foreach my $bed(@$pckBed){
                print $spnH join("\t",
                    (map {$$prp[$c{$_}]} qw(grpId sample)),
                    $sample, $pairType, @$bed[2..3], $$bed[0]+$maxLineAll
                ),"\n"; 
            }
            $maxLineAll += $maxLine + 2;
        }    
    }    
}
sub packSpans {  # minimize vertical display by packing spans
    my ($spns) = @_; # expects $$spns{$p1}{$p2} = [ [], ... ]
    my ($baseLine, $maxLine, %lineMaxs, @pckFeats) = (0, 0);
    foreach my $p1(sort {$a <=> $b} keys %$spns){
        foreach my $p2(sort {$a <=> $b} keys %{$$spns{$p1}}){
            foreach my $feat(@{$$spns{$p1}{$p2}}){
                my $line = $baseLine;
                while(defined $lineMaxs{$line} and $lineMaxs{$line} >= $p1){ $line++ }
                $maxLine >= $line or $maxLine = $line;
                $lineMaxs{$line} = $p2;
                unshift @$feat, $line; # put assigned line for this feature in array index 0
                push @pckFeats, $feat; # assemble an array of packed features
            }  
        }   
    }
    return (\@pckFeats, $maxLine); # return the packed features and maximum assigned line
}

# print SNP coverage
sub printSNPs {
    my ($prp, $plotRegion) = @_;    
    foreach my $sample(keys %allSamples){
    #foreach my $sample($$prp[$c{sample}]){   
        my $genSnpsFile = "$genSnpsDir/svtools.genotype.snps.$sample.bgz";
        -e $genSnpsFile or next;
        open my $tbxH, "-|", "tabix $genSnpsFile $plotRegion" or
            die "could not open tabix stream: tabix $genSnpsFile $plotRegion\n";      
        while (my $snp = <$tbxH>) {
            chomp $snp;
            my @snp = split("\t", $snp);
            print $snpH join("\t",
                (map {$$prp[$c{$_}]} qw(grpId sample)),
                $sample,
                (map {$snp[$gntCols{$_}]} qw (end nRef nAlt refBase altBase))
            ),"\n";
        }        
        close $tbxH;
    }         
}

# highlight genome gaps and seg dups
sub printGaps {
    my ($prp, $plotRegion) = @_;
    print $gapH join("\t", # make sure the file is not empty
        (map {$$prp[$c{$_}]} qw(grpId)),
        -1, -1, 'black'
    ), "\n"; 
    $gapFile and printGaps_($gapFile, 'darkgrey', $prp, $plotRegion);
    $segDupFile and printGaps_($segDupFile, 'lightblue', $prp, $plotRegion);
}
sub printGaps_ {
    my ($file, $col, $prp, $plotRegion) = @_;
    open my $tbxH, "-|", "tabix $file $plotRegion" or
        die "could not open tabix stream: tabix $file $plotRegion\n";    
    while (my $bed = <$tbxH>) {
        chomp $bed;
        my @bed = split("\t", $bed);
        print $gapH join("\t",
            (map {$$prp[$c{$_}]} qw(grpId)),
            $bed[1], $bed[2], $col
        ), "\n"; 
    }    
    close $tbxH;
}

1;

#Col	Field	Description
#1	QNAME	Query template/pair NAME
#2	FLAG	bitwise FLAG
#3	RNAME	Reference sequence NAME
#4	POS	1-based leftmost POSition/coordinate of clipped sequence
#5	MAPQ	MAPping Quality (Phred-scaled)
#6	CIAGR	extended CIGAR string
#7	MRNM	Mate Reference sequence NaMe (?=? if same as RNAME)
#8	MPOS	1-based Mate POSistion
#9	TLEN	inferred Template LENgth (insert size)
#10	SEQ	query SEQuence on the same strand as the reference
#11	QUAL	query QUALity (ASCII-33 gives the Phred base quality)
#12+	OPT	variable OPTional fields in the format TAG:VTYPE:VALUE
#Each bit in the FLAG field is defined as:
#
#Flag	Chr	Description
#0x001	p	the read is paired in sequencing
#0x002	P	the read is mapped in a proper pair
#0x004	u	the query sequence itself is unmapped
#0x008	U	the mate is unmapped
#0x010	r	strand of the query (1 for reverse)
#0x020	R	strand of the mate
#0x040	1	the read is the first read in a pair
#0x080	2	the read is the second read in a pair
#0x100	s	the alignment is not primary
#0x200	f	the read fails platform/vendor quality checks
#0x400	d	the read is either a PCR or an optical duplicate
#0x800      supplementary alignment
